### PR TITLE
build tables for url

### LINKED ISSUE #146: build tables for url
URL does not build or use the lookup tables. It needs to do both. It might be worth looking into pushing the copied strings logic deeper into the implementation first.

### PR DESCRIPTION
**Pull request by alandefreitas**
Work in progress. We're building the tables but not using the tables yet.

fix #146

### CODE REVIEW

File: include/boost/url/detail/config.hpp [hunk 1]
```diff
@@ -12,7 +12,8 @@
 
 #include <boost/config.hpp>
 #include <boost/config/workaround.hpp>
-#include <limits.h>
+#include <climits>
+#include <cstdint>
```
Comment (vinniefalco) [inline include/boost/url/detail/config.hpp:16]: I really try to keep the number of includes in this file down to a minimum, because it is included everywhere. That's why I included limits.h instead of cstdint and climits.

File: include/boost/url/impl/url.ipp [hunk 2]
```diff
@@ -2271,63 +2274,64 @@ void
 url::
 build_tab() noexcept
```
Comment (vinniefalco) [inline include/boost/url/impl/url.ipp:2275]: this should follow the fields_view_base interface of `write_table`: 
https://github.com/CPPAlliance/http_proto/blob/3168cbadff093a837f45a4414656951d154f2486/include/boost/http_proto/impl/fields_view_base.ipp#L263

File: include/boost/url/impl/url.ipp [hunk 3]
```diff
@@ -2271,63 +2274,64 @@ void
...
 {
-#if 0
-    // path
-    if(nseg_ > 1)
+    BOOST_ASSERT(s_);
+
+    if (nseg_ != 0)
```
Comment (vinniefalco) [inline include/boost/url/impl/url.ipp:2279]: we already know the beginning of the first segment (`offset(id_path)`) and the end of the last segment (`offset(id_query)`). So the table doesn't need to store these.

File: include/boost/url/impl/url.ipp [hunk 4]
```diff
@@ -2351,6 +2355,21 @@ ensure_space(
...
+        sizeof(detail::segments_table_entry);
+    new_cap = AS * ((new_cap + AS - 1) / AS) +
+           (nseg * SS);
+
+    static constexpr auto AP =
+        alignof(detail::params_table_entry);
+    static constexpr auto SP =
+        sizeof(detail::params_table_entry);
+    new_cap = AP * ((new_cap + AP - 1) / AP) +
+           (nparam * SP);
```
Comment (vinniefalco) [inline include/boost/url/impl/url.ipp:2371]: The problem here is we have two tables back to back. That means that any changes in size of the second table are going to cause memmove of the entire first table. I was thinking instead that we have just one table with enough space for the larger of the two sizes, where each element stores a param and a segment offset.

File: include/boost/url/impl/url.ipp [hunk 5]
```diff
@@ -2366,8 +2385,17 @@ ensure_space(
         }
         if( new_cap < n)
             new_cap = n;
+        std::size_t old_cap = cap_;
         s = allocate(new_cap);
         std::memcpy(s, s_, size());
+        auto sp = s_ + old_cap;
+        auto st = detail::segments_table(sp);
+        st.copy(s + cap_, nseg_);
```
Comment (vinniefalco) [inline include/boost/url/impl/url.ipp:2393]: well you are adding `old_cap` but then using `cap_` why not just leave out the local variable copy and just use `cap_`?

File: include/boost/url/impl/url.ipp (hunk 2 above)
Comment (alandefreitas) [reply include/boost/url/impl/url.ipp:2275]: Oh... OK. I attempted to replicate the logic in `parser::parse_fields`.

Comment (alandefreitas) [reply include/boost/url/impl/url.ipp:2275]: I'm not sure I know what to do here. 

1) Should we keep `build_tab` and make it call some kind of `write_table` functions somewhere else? Where? What would `buf_len_` represent in that case?
2) Or should we just reuse the logic from `write_table` in `build_tab`? If so, I'm assuming the equivalent of `fields_view_base::buf_len_` would be `url::cap_`, but the initial condition `buf_len_ > 0` (or `cap_ > 0`) will always be true, because we allocate before writing the table in the `url`.

File: include/boost/url/impl/url.ipp (hunk 3 above)
Comment (alandefreitas) [reply include/boost/url/impl/url.ipp:2279]: Yes. I saw that logic for fields. 

Are they also always at the beginning for segments and query parameters? 

I thought these views (segments and params) implemented more complex logic to get rid of edge case prefixes and suffixes so I thought it wasn't worth reimplementing this logic elsewhere.

File: include/boost/url/impl/url.ipp (hunk 4 above)
Comment (alandefreitas) [reply include/boost/url/impl/url.ipp:2371]: > I was thinking instead that we have just one table with enough space for the larger of the two sizes, where each element stores a param and a segment offset.

Mmm... This is possible. In that case, we would need a new name for the `<component>_table_entry`, `<component>_table`, and `const_<component>_table`. 

It's also a trade-off but I'm not sure of all the implications:

- From a perspective of memory, we are obviously always spending more memory, unless `nseg == nparam`, which should be quite rare.
- From a perspective of computational cost, I have the impression the merged table is absorbing the cost of manually moving elements into any operation changing any of the tables, which is more costly than memmoving only one of the tables.
- From the perspective of code complexity, I'm not sure yet, but I have the impression we're going to have lots of ifs and elses to do what memmove would.

File: include/boost/url/impl/url.ipp (hunk 5 above)
Comment (alandefreitas) [reply include/boost/url/impl/url.ipp:2393]: Nice. Yes, it's possible to avoid the `old_cap` if we get the references to the tables first.

Before changing this logic, I stored the `old_cap` before because `s = allocate(new_cap);` would invalidate the old `cap_`.

File: include/boost/url/impl/url.ipp (hunk 4 above)
Comment (vinniefalco) [reply include/boost/url/impl/url.ipp:2371]: Well I will leave the decision up to you

File: include/boost/url/impl/url.ipp (hunk 3 above)
Comment (vinniefalco) [reply include/boost/url/impl/url.ipp:2279]: I don't think there are any edge cases, but I could be wrong. Its up to you.

File: include/boost/url/impl/url.ipp (hunk 2 above)
Comment (vinniefalco) [reply include/boost/url/impl/url.ipp:2275]: What I'm saying is that having the function `url_view::write_table( void* dest )` seems like it would be a good primitive, based on experience in HTTP fields. However, unlike fields we have two tables. Yes `buf_len_` in fields is the same as `cap_` in URL (maybe I will rename it heh).

I will leave the decision on implementation interface up to you.

File: include/boost/url/impl/url.ipp (hunk 3 above)
Comment (alandefreitas) [reply include/boost/url/impl/url.ipp:2279]: There are some considerations that make it a little more complex than fields because the first element is not at index `0`.

Like

https://github.com/CPPAlliance/url/blob/fbc51ed52412d1421767effa92951ff417c1691f/include/boost/url/detail/path.hpp#L23

and

https://github.com/CPPAlliance/url/blob/fbc51ed52412d1421767effa92951ff417c1691f/include/boost/url/impl/url_view.ipp#L402

This logic is not complex. It's just more paths to cover because inferring the first `n` elements requires us to replicate this logic elsewhere to reconstruct these views to find the element.

But I'll start with the simpler logic and adapt on top of that.

File: include/boost/url/detail/config.hpp [hunk 6]
```diff
@@ -84,4 +85,14 @@
 # error Unknown or unsupported architecture, please open an issue
 #endif
 
+namespace boost {
+
+namespace urls {
+
+using off_t = ::uint16_t; // private
```
Comment (vinniefalco) [inline include/boost/url/detail/config.hpp:92]: what's wrong with unsigned short?

File: include/boost/url/detail/params_table.hpp [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/CPPAlliance/url
+//
+
+#ifndef BOOST_URL_DETAIL_PARAMS_TABLE_HPP
+#define BOOST_URL_DETAIL_PARAMS_TABLE_HPP
+
+#include <boost/url/string_view.hpp>
+#include <boost/assert.hpp>
```
Comment (vinniefalco) [inline include/boost/url/detail/params_table.hpp:15]: do you need this include?

File: include/boost/url/detail/params_table.hpp [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+//------------------------------------------------
+
+// params array stored at the
+// end of allocated url string data
+
+struct params_table_entry
+{
+    off_t kp;   // key pos
+    off_t kn;   // key size
+    off_t vp;   // value pos
```
Comment (vinniefalco) [inline include/boost/url/detail/params_table.hpp:33]: well if you know the key size then you also know the value pos, right? and if you know the position of the next key then that gives you the value length?

File: include/boost/url/detail/config.hpp (hunk 6 above)
Comment (alandefreitas) [reply include/boost/url/detail/config.hpp:92]: I was hoping you would comment on that. I was going to use `size_t` or `uint32_t` directly on the table class, then I got this pattern from http_proto. What I don't get is why `uint16_t` is OK for fields. Why is it?

Comment (vinniefalco) [reply include/boost/url/detail/config.hpp]: don't you think 64kb is enough for an HTTP header?

Comment (alandefreitas) [reply include/boost/url/detail/config.hpp]: For the median HTTP header, yes. For the worst-case HTTP header according to the spec, I don't know.

For the median URL header, yes. For the worst-case URL according to the spec, no because of a [technicality](https://stackoverflow.com/questions/417142/what-is-the-maximum-length-of-a-url-in-different-browsers).

File: include/boost/url/impl/url.ipp [hunk 9]
```diff
@@ -1515,49 +1519,25 @@ param(
...
+    detail::const_params_table t(param_tab_begin());
+    string_view q = encoded_query();
+    auto const& e = t[i];
+    const char* start = q.data() + e.kp - 1;
+    const char* end = s_ + offset(id_frag);
+    if (i != nparam_ - 1)
+        end = q.data() + t[i + 1].kp - 1;
+
+    std::size_t param_pos = start - s_;
+    std::size_t param_n = end - start;
```
Comment (vinniefalco) [inline include/boost/url/impl/url.ipp:1537]: I would use auto

File: include/boost/url/detail/segments_table.hpp [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+    {
+    }
+
+    segments_table_entry const&
+    operator[](
+        std::size_t i) const noexcept
+    {
+        return p_[-1 * (
+            static_cast<
+                long>(i) + 1)];
```
Comment (vinniefalco) [inline include/boost/url/detail/segments_table.hpp:53]: converting size_t to long is narrowing when size_t is 64-bits

File: include/boost/url/impl/url.ipp (hunk 9 above)
Comment (alandefreitas) [reply include/boost/url/impl/url.ipp:1537]: `auto param_pos = start - s_;` would become `long param_pos = = start - s_;` and fail in the `raw_param` constructor.

```
Non-constant-expression cannot be narrowed from type 'long' to 'std::size_t' (aka 'unsigned long') in initializer list
```

**Comment by alandefreitas** [comment]
The problem with segment prefixes and suffixes has been fixed. 

To do list:

- Improve coverage
- Add documentation about the malleable prefixes and suffixes
- Add a few relevant edge cases to the tests

**Comment by alandefreitas** [comment]
Closing this PR as there is not enough user demand at this point.
